### ISSUE TITLE
Failure trying to install ghc head

### ISSUE THREAD

**Issue by lf-** [haskell/actions#137]
https://github.com/haskell/actions/blob/main/setup/src/installer.ts#L291

Failing run: https://github.com/haskell/haddock/actions/runs/3653393844/jobs/6172813208

some log:

```
   Attempting to install ghc head using ghcup
  /opt/hostedtoolcache/ghcup/0.1.18.0/x64/ghcup install ghc -u https://gitlab.haskell.org/ghc/ghc/-/jobs/artifacts/master/raw/ghc-x86_64-deb9-linux-integer-simple.tar.xz?job=validate-x86_64-linux-deb9-integer-simple head
  [ Info  ] downloading: https://raw.githubusercontent.com/haskell/ghcup-metadata/master/ghcup-0.0.7.yaml as file /home/runner/.ghcup/cache/ghcup-0.0.7.yaml
    % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                   Dload  Upload   Total   Spent    Left  Speed
  
    0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
    0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  [ Warn  ] Could not get download info, trying cached version (this may not be recent!)
  [ ...   ] If this problem persists, consider switching downloader via: 
  [ ...   ]     ghcup config set downloader Wget
  [ Warn  ] New cabal version available. To upgrade, run 'ghcup install cabal 3.8.1.0'
  [ Info  ] downloading: https://gitlab.haskell.org/ghc/ghc/-/jobs/artifacts/master/raw/ghc-x86_64-deb9-linux-integer-simple.tar.xz?job=validate-x86_64-linux-deb9-integer-simple as file /home/runner/.ghcup/tmp/ghcup-3f85036592d7f958/ghc-x86_64-deb9-linux-integer-simple.tar.xz
    % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                   Dload  Upload   Total   Spent    Left  Speed
  
    0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
    0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  curl: (22) The requested URL returned error: 404
  [ Error ] Download failed: Process "curl" with arguments ["-fL", "-o",
  [ ...   ]                                                 "/home/runner/.ghcup/tmp/ghcup-3f85036592d7f958/ghc-x86_64-deb9-linux-integer-simple.tar.xz.tmp",
  [ ...   ]                                                 "https://gitlab.haskell.org/ghc/ghc/-/jobs/artifacts/master/raw/ghc-x86_64-deb9-linux-integer-simple.tar.xz?job=validate-x86_64-linux-deb9-integer-simple"] failed with exit code 22.
  [ Error ] Also check the logs in /home/runner/.ghcup/logs
Error: All install methods for ghc head failed
```

**Comment by lf-** [comment]
hypothesis: ghc changed their artefact names, possibly?

**Comment by johnckolesar** [comment]
I'm experiencing a similar issue with a repository of my own:

> Run haskell/actions/setup@v2
Preparing to setup a Haskell environment
Resolved 3.0 to 3.0.0.0
Preparing ghc environment
Installing ghc version [8](https://github.com/BillHallahan/G2/actions/runs/3681627706/jobs/6228506601#step:3:9).2.2
Error: All install methods for ghc 8.2.2 failed

This comes from a build failure for a branch pull request.  When I first opened this pull request, I didn't get any errors like this.  If this is a problem on GHC's end, then it must be the result of some change that happened within the past two weeks.

**Comment by BillHallahan** [comment]
To add a bit of information, a failing CI run for the repository @johnckolesar is referring to is here:
https://github.com/BillHallahan/G2/actions/runs/3681627706/jobs/6228506601
The end result is the same (ghc head fails to install) but we get a gmake error, rather than the curl error that @lf- gets:
 
```
[ Info  ] Installing GHC (this may take a while)
...
  [ Error ] Process "gmake" with arguments ["DESTDIR=/home/runner/.ghcup/tmp/ghcup-0c8e105d1542c119",
[48](https://github.com/BillHallahan/G2/actions/runs/3681627706/jobs/6228506601#step:3:50)
  [ ...   ]                                 "install"] failed with exit code 2.
[49](https://github.com/BillHallahan/G2/actions/runs/3681627706/jobs/6228506601#step:3:51)
  [ Error ] Also check the logs in /home/runner/.ghcup/logs
[50](https://github.com/BillHallahan/G2/actions/runs/3681627706/jobs/6228506601#step:3:52)
  Attempting to install ghc 8.2.2 using apt-get
```
Running on MacOS rather than Ubuntu seems to be a (temporary) fix for the problem:
https://github.com/BillHallahan/G2/pull/169

**Comment by lf-** [comment]
ah yes of course it's that, there's other breakage with ubuntu-latest changing too

**Comment by andreasabel** [comment]
@BillHallahan 
> Attempting to install ghc 8.2.2 using apt-get

@johnckolesar 
> Error: All install methods for ghc 8.2.2 failed

This is not about GHC head, but it is actually:
- #135

Discussion pertaining to GHC 8.2, 8.0 on `ubuntu-22.04` should happen in #135, not here.

**Comment by andreasabel** [comment]
@lf- Thanks for the report.

@lf- wrote
> [`main`/setup/src/installer.ts#L291](https://github.com/haskell/actions/blob/main/setup/src/installer.ts?rgh-link-date=2022-12-09T02%3A09%3A05Z#L291)

You didn't put a permalink here, so this might point to a different line you intended.  (Please always link to github content with permalinks, you can e.g. get it by typing `y`.)

> Failing run: [haskell/haddock/actions/runs/3653393844/jobs/6172813208](https://github.com/haskell/haddock/actions/runs/3653393844/jobs/6172813208)

Could you provide more information to why `setup` is goint wrong here?   E.g., why is it not a problem upstream, in ghcup?   Does `setup` need to update its `ghcup` version?  Etc.

**Comment by lf-** [comment]
Apologies, here: https://github.com/haskell/actions/blob/3c89646d961d828e78fdd123db1dd64cca53a7a3/setup/src/installer.ts#L312

The reason I believe it's setup doing it is that it's 404ing on a url that seems to come from setup.

**Comment by andreasabel** [comment]
Thanks.  I guess this would also affect ghcup (ping @hasufell), at least the help text:
```console
$ ghcup install ghc --help
...
  # install GHC head fedora bindist
  ghcup install ghc -u https://gitlab.haskell.org/api/v4/projects/1/jobs/artifacts/master/raw/ghc-x86_64-fedora27-linux.tar.xz?job=validate-x86_64-linux-fedora27 head
```
The URL given here is also a 404.

@mpickering Matthew, what would the new URLs be? (Pinging you because you published https://mpickering.github.io/posts/2019-06-11-ghc-artefact.html.)

**Comment by mpickering** [comment]
@andreasabel You probably want to use `fedora33` bindists instead.

```
https://gitlab.haskell.org/api/v4/projects/1/jobs/artifacts/master/raw/ghc-x86_64-linux-fedora33-release.tar.xz?job=x86_64-linux-fedora33-release
```

**Comment by hasufell** [comment]
> @andreasabel You probably want to use `fedora33` bindists instead.
> 
> ```
> https://gitlab.haskell.org/api/v4/projects/1/jobs/artifacts/master/raw/ghc-x86_64-linux-fedora33-release.tar.xz?job=x86_64-linux-fedora33-release
> ```

Maybe not, because it will result in glibc errors on many ubuntu versions. You can try to pick deb9 or centos7.

**Comment by andreasabel** [comment]
@mpickering @hasufell: Thanks for chiming in.
- exists: https://gitlab.haskell.org/api/v4/projects/1/jobs/artifacts/master/raw/ghc-x86_64-linux-fedora33-release.tar.xz?job=x86_64-linux-fedora33-release
- exists not: https://gitlab.haskell.org/api/v4/projects/1/jobs/artifacts/master/raw/ghc-x86_64-linux-deb9-release.tar.xz?job=x86_64-linux-deb9-release
- exists not reprise: https://gitlab.haskell.org/api/v4/projects/1/jobs/artifacts/master/raw/ghc-x86_64-linux-centos7-release.tar.xz?job=x86_64-linux-centos7-release

Where do I get these magic links from (the valid ones)?  These URLs are not very exploreable, prefixes of these give me uniformly 404.

**Comment by andreasabel** [comment]
@mpickering : Please point me to a list of the GHC HEAD artifacts URLs.

**Comment by mpickering** [comment]
Which bindists do you actually want @andreasabel? The link I suggested for fedora33 works? 

If I read your URLs it seems you want `deb9-release` from master branch.. which isn't a job which exists and also `centos7-release` from master branch, also something which doesn't exist. 

On the master branch we run a "validate" pipeline on each commit and a "nightly" pipeline once a day. Neither of which produce deb9-release or centos7-release bindists. 

Example validate: https://gitlab.haskell.org/ghc/ghc/-/pipelines/60392
Example nightly: https://gitlab.haskell.org/ghc/ghc/-/pipelines/60384

Perhaps instead you are interested in something like the "discover-tarball" script which we use on head.hackage (https://gitlab.haskell.org/ghc/head.hackage/-/blob/master/ci/discover_tarball.sh).

**Comment by Bodigrim** [comment]
@andreasabel you can have any bindist from `full-build` section of pipelines. 

How about Alpine bindist? It should be resilent to `glibc` / `libtinfo5` issues.

**Comment by hasufell** [comment]
> How about Alpine bindist?

Is usually broken, because it's fully static and has all sorts of linking problems.

**Comment by Bodigrim** [comment]
Yeah, it used to be, but it seems that recently Alpine job is green.

**Comment by hasufell** [comment]
> Yeah, it used to be, but it seems that recently Alpine job is green.

Yes, it's green and produces an unusable bindist. Alpine bindists for latest GHCs are practically broken for end users.

**Comment by andreasabel** [comment]
@mpickering wrote:
> Perhaps instead you are interested in something like the "discover-tarball" script which we use on head.hackage ([gitlab.haskell.org/ghc/head.hackage/-/blob/master/ci/discover_tarball.sh](https://gitlab.haskell.org/ghc/head.hackage/-/blob/master/ci/discover_tarball.sh)).

Yes, I would like to have a robust way to discover the GHC HEAD binaries.  I had a look at this script, but it depends on specific environment variables to be set.

**Comment by andreasabel** [comment]
@mpickering @Bodigrim @hasufell : Thanks for your help!

From the discussion I got the impression that there is currently no robust way to install GHC head from binaries.
I thus went ahead and un-advertised the feature in the README.  So, currently installing head isn't supported.

That said, I am not closing the door on this, more information that leads to a fix is always welcome, as well as PRs that close this issue.

**Comment by andreasabel** [comment]
Solved by #279.
